File: articles-api-typescript/src/routes/articles.ts
import { Router, Request, Response } from "express";
import { DataSource } from "typeorm";
import { ArticleService } from "../services/ArticleService";
import { AuthenticatedRequest, requireAuth } from "../middleware/auth";
import { CreateArticleRequest, UpdateArticleRequest } from "../types";

export function createArticlesRouter(dataSource: DataSource): Router {
  const router = Router();
  const articleService = new ArticleService(dataSource);

  /**
   * GET /api/articles
   * List articles with optional filters: tag, author, favorited, offset, limit.
   * Mirrors Java ArticlesApi.getArticles()
   */
  router.get("/", async (req: Request, res: Response) => {
    try {
      const offset = parseInt(req.query.offset as string) || 0;
      const limit = parseInt(req.query.limit as string) || 20;
      const tag = req.query.tag as string | undefined;
      const author = req.query.author as string | undefined;
      const favorited = req.query.favorited as string | undefined;
      const currentUser = (req as AuthenticatedRequest).user;

      const { articles, count } = await articleService.listArticles(
        offset,
        limit,
        tag,
        author,
        favorited
      );

      const enriched = await articleService.enrichArticles(
        articles,
        currentUser
      );

      res.json({ articles: enriched, articlesCount: count });
    } catch (err) {
      res.status(500).json({ errors: { body: ["internal server error"] } });
    }
  });

  /**
   * GET /api/articles/feed
   * Articles from followed authors (requires auth).
   * Mirrors Java ArticlesApi.getFeed()
   */
  router.get("/feed", requireAuth, async (req: Request, res: Response) => {
    try {
      const offset = parseInt(req.query.offset as string) || 0;
      const limit = parseInt(req.query.limit as string) || 20;
      const currentUser = (req as AuthenticatedRequest).user!;

      const { articles, count } = await articleService.getFeed(
        currentUser.id,
        offset,
        limit
      );

      const enriched = await articleService.enrichArticles(
        articles,
        currentUser
      );

      res.json({ articles: enriched, articlesCount: count });
    } catch (err) {
      res.status(500).json({ errors: { body: ["internal server error"] } });
    }
  });

  /**
   * POST /api/articles
   * Create article (requires auth).
   * Mirrors Java ArticlesApi.createArticle()
   */
  router.post("/", requireAuth, async (req: Request, res: Response) => {
    try {
      const currentUser = (req as AuthenticatedRequest).user!;
      const body = req.body as CreateArticleRequest;

      if (
        !body.article ||
        !body.article.title ||
        !body.article.description ||
        !body.article.body
      ) {
        res.status(422).json({
          errors: { body: ["title, description, and body are required"] },
        });
        return;
      }

      const article = await articleService.createArticle(
        body.article.title,
        body.article.description,
        body.article.body,
        body.article.tagList || [],
        currentUser.id
      );

      const enriched = await articleService.enrichArticle(
        article,
        currentUser
      );

      res.json({ article: enriched });
    } catch (err) {
      res.status(500).json({ errors: { body: ["internal server error"] } });
    }
  });

  /**
   * GET /api/articles/:slug
   * Get single article by slug.
   * Mirrors Java ArticleApi.article()
   */
  router.get("/:slug", async (req: Request, res: Response) => {
    try {
      const currentUser = (req as AuthenticatedRequest).user;
      const article = await articleService.findBySlug(req.params.slug);

      if (!article) {
        res.status(404).json({ errors: { body: ["article not found"] } });
        return;
      }

      const enriched = await articleService.enrichArticle(
        article,
        currentUser
      );

      res.json({ article: enriched });
    } catch (err) {
      res.status(500).json({ errors: { body: ["internal server error"] } });
    }
  });

  /**
   * PUT /api/articles/:slug
   * Update article (requires auth, author only).
   * Mirrors Java ArticleApi.updateArticle()
   */
  router.put("/:slug", requireAuth, async (req: Request, res: Response) => {
    try {
      const currentUser = (req as AuthenticatedRequest).user!;
      const article = await articleService.findBySlug(req.params.slug);

      if (!article) {
        res.status(404).json({ errors: { body: ["article not found"] } });
        return;
      }

      if (article.userId !== currentUser.id) {
        res.status(403).json({ errors: { body: ["not authorized"] } });
        return;
      }

      const body = req.body as UpdateArticleRequest;
      const updated = await articleService.updateArticle(
        article,
        body.article?.title,
        body.article?.description,
        body.article?.body
      );

      const enriched = await articleService.enrichArticle(
        updated,
        currentUser
      );

      res.json({ article: enriched });
    } catch (err) {
      res.status(500).json({ errors: { body: ["internal server error"] } });
    }
  });

  /**
   * DELETE /api/articles/:slug
   * Delete article (requires auth, author only).
   * Mirrors Java ArticleApi.deleteArticle()
   */
  router.delete(
    "/:slug",
    requireAuth,
    async (req: Request, res: Response) => {
      try {
        const currentUser = (req as AuthenticatedRequest).user!;
        const article = await articleService.findBySlug(req.params.slug);

        if (!article) {
          res.status(404).json({ errors: { body: ["article not found"] } });
          return;
        }

        if (article.userId !== currentUser.id) {
          res.status(403).json({ errors: { body: ["not authorized"] } });
          return;
        }

        await articleService.deleteArticle(article);
        res.status(204).send();
      } catch (err) {
        res.status(500).json({ errors: { body: ["internal server error"] } });
      }
    }
  );

  return router;
}

File: python-fastapi/app/routers/articles.py
from typing import Optional

from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy.orm import Session

from app.database import get_db
from app.models.models import Article
from app.models.models import User as UserModel
from app.schemas.article import (
    MultipleArticlesResponse,
    NewArticleWrapper,
    SingleArticleResponse,
    UpdateArticleWrapper,
)
from app.services.article_service import (
    create_article,
    delete_article,
    get_article_by_slug,
    get_feed,
    list_articles,
    update_article,
)
from app.services.auth import get_current_user_optional, get_current_user_required

router = APIRouter(prefix="/api/articles", tags=["Articles"])


@router.get("", response_model=MultipleArticlesResponse)
def get_articles(
    tag: Optional[str] = Query(None),
    author: Optional[str] = Query(None),
    favorited: Optional[str] = Query(None),
    offset: int = Query(0),
    limit: int = Query(20),
    db: Session = Depends(get_db),
    current_user: Optional[UserModel] = Depends(get_current_user_optional),
):
    articles, total = list_articles(
        db, current_user, tag=tag, author=author, favorited=favorited, offset=offset, limit=limit
    )
    return MultipleArticlesResponse(articles=articles, articlesCount=total)


@router.get("/feed", response_model=MultipleArticlesResponse)
def get_article_feed(
    offset: int = Query(0),
    limit: int = Query(20),
    db: Session = Depends(get_db),
    current_user: UserModel = Depends(get_current_user_required),
):
    articles, total = get_feed(db, current_user, offset=offset, limit=limit)
    return MultipleArticlesResponse(articles=articles, articlesCount=total)


@router.get("/{slug}", response_model=SingleArticleResponse)
def get_single_article(
    slug: str,
    db: Session = Depends(get_db),
    current_user: Optional[UserModel] = Depends(get_current_user_optional),
):
    article_data = get_article_by_slug(slug, current_user, db)
    if article_data is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Article not found")
    return SingleArticleResponse(article=article_data)


@router.post("", response_model=SingleArticleResponse, status_code=status.HTTP_200_OK)
def create_new_article(
    payload: NewArticleWrapper,
    db: Session = Depends(get_db),
    current_user: UserModel = Depends(get_current_user_required),
):
    try:
        article_data = create_article(
            db,
            current_user,
            title=payload.article.title,
            description=payload.article.description,
            body=payload.article.body,
            tag_list=payload.article.tagList,
        )
    except ValueError as e:
        raise HTTPException(status_code=status.HTTP_422_UNPROCESSABLE_ENTITY, detail=str(e))
    return SingleArticleResponse(article=article_data)


@router.put("/{slug}", response_model=SingleArticleResponse)
def update_existing_article(
    slug: str,
    payload: UpdateArticleWrapper,
    db: Session = Depends(get_db),
    current_user: UserModel = Depends(get_current_user_required),
):
    article = db.query(Article).filter(Article.slug == slug).first()
    if article is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Article not found")
    if article.user_id != current_user.id:
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="Not authorized")
    article_data = update_article(
        db,
        article,
        current_user,
        title=payload.article.title,
        description=payload.article.description,
        body=payload.article.body,
    )
    return SingleArticleResponse(article=article_data)


@router.delete("/{slug}", status_code=status.HTTP_204_NO_CONTENT)
def delete_existing_article(
    slug: str,
    db: Session = Depends(get_db),
    current_user: UserModel = Depends(get_current_user_required),
):
    article = db.query(Article).filter(Article.slug == slug).first()
    if article is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Article not found")
    if article.user_id != current_user.id:
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="Not authorized")
    delete_article(db, article)
